### PR TITLE
Omps integration

### PR DESCRIPTION
**Pull request by MartinBasti**
Missing parts:
- [x] unittest
- [x] OMPS query retries
- [x] Set timeouts for queries

documentation update: https://github.com/containerbuildsystem/osbs-docs/pull/85

Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates

### CODE REVIEW

File: atomic_reactor/omps_util.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+"""
+Copyright (c) 2019 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
```
Comment (athos-ribeiro) [inline atomic_reactor/omps_util.py]: Should this module be implemented in OMPS upstream and imported here?

File: atomic_reactor/plugins/exit_koji_import.py [hunk 2]
```diff
@@ -224,17 +228,27 @@ def set_go_metadata(self, extra):
             extra['image']['go'] = go
 
     def set_operators_metadata(self, extra, worker_metadatas):
+        # update push plugin and uploaded manifests file independently as push plugin may fail
+        op_push_res = self.workflow.postbuild_results.get(PLUGIN_PUSH_OPERATOR_MANIFESTS_KEY)
+        if op_push_res:
+            extra.update({
+                "operator_manifests": {
```
Comment (athos-ribeiro) [inline atomic_reactor/plugins/exit_koji_import.py]: should this go in the uploaded manifests metadata though?

File: atomic_reactor/plugins/post_push_operator_manifest.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+        workdir = self.workflow.source.get_build_file_path()[1]
+        dest_dir = os.path.join(workdir, self.DOWNLOAD_DIR)
+        dest_path = os.path.join(dest_dir, dest_filename)
+        if not os.path.exists(dest_dir):
+            os.makedirs(dest_dir)
+
+        req_session = get_retrying_requests_session()
+        request = req_session.get(url, stream=True,
+                                  verify=False)  # TODO configurable
```
Comment (athos-ribeiro) [inline atomic_reactor/plugins/post_push_operator_manifest.py]: Just a note to check this TODO before final review

File: atomic_reactor/plugins/post_push_operator_manifest.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+        if not self.should_run():
+            return
+
+        omps_config = get_omps_config(self.workflow)
+        omps = OMPS.from_config(omps_config)
+
+        operator_manifest_url = self.get_koji_operator_manifest_url()
+        operator_manifests_path = self.download_file(
+            operator_manifest_url, 'operator-manifests.zip')
```
Comment (athos-ribeiro) [inline atomic_reactor/plugins/post_push_operator_manifest.py]: There is a constant for this filename: OPERATOR_MANIFESTS_ARCHIVE

File: atomic_reactor/omps_util.py (hunk 1 above)
Comment (MartinBasti) [reply atomic_reactor/omps_util.py]: I'm using OSBS specific retries. I'd not complicate things with 3rd party dependencies. OMPS support only py3.6+

File: atomic_reactor/plugins/exit_koji_import.py (hunk 2 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/exit_koji_import.py]: IIRC we had discussion about this in design phase and agreed on this

File: atomic_reactor/plugins/post_push_operator_manifest.py (hunk 3 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/post_push_operator_manifest.py]: I made it configurable

File: atomic_reactor/omps_util.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                "OMPS service request failed with error: {err}: {msg}".format(
+                    err=error,
+                    msg=msg
+                ),
+                status_code=response.status_code,
+                response=response_json
+            )
+
+    def push_archive(self, fb):
+        """Puhs operator manifest archive to appregistry via OMPS service
```
Comment (rcerven) [inline atomic_reactor/omps_util.py]: typo "Puhs"

**Comment by rcerven** [comment]
looks great

File: atomic_reactor/plugins/post_push_operator_manifest.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                result = omps.push_archive(fb)
+        except OMPSError as e:
+            msg = "Failed to push operator manifests: {}".format(e)
+            self.log.error(msg)
+            raise RuntimeError(msg)
+
+        try:
+            os.remove(operator_manifests_path)
+        except OSError as e:
+            self.log.warning("Cleanup of the downloaded archive failed: %s", e)
```
Comment (athos-ribeiro) [inline atomic_reactor/plugins/post_push_operator_manifest.py]: This is missing coverage from the unit tests. Could we have a test case for this one?

Comment (MartinBasti) [reply atomic_reactor/plugins/post_push_operator_manifest.py]: Do we really need test for this?

File: atomic_reactor/omps_util.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    @classmethod
+    def from_config(cls, config):
+        """Initialize instance from reactor config map"""
+        with open(os.path.join(config['omps_secret_dir'], 'token'), 'r') as f:
+            token = f.read().strip()
+        return cls(
+            config['omps_url'],
+            config['omps_namespace'],
+            token
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/omps_util.py]: nice catch

File: docs/plugins.md [hunk 8]
```diff
@@ -189,6 +189,9 @@ These are run after buildstep plugin has successfully finished.
  * **koji_upload**
    * Status: enabled
    * The 'docker save' output, build logs, and operator manifests are uploaded to Koji. The metadata is returned to be used by the store_metadata_osv3 plugin.  That plugin will use a ConfigMap object to store it for the orchestrator to retrieve it.  It will replace koji_promote when enabled.
+ * **push_operator_manifests**
+   * Status: enabled
+   * When specified through the com.redhat.delivery.appregistry Dockerfile label, plugin uploads manifests extracted by *export_operator_manifests* into app registry specified in configuration.
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply docs/plugins.md]: push plugin checks for that label too, I'll extend it about information about config

**Comment by MartinBasti** [comment]
updated (except the missing os.remove coverage)

**Comment by athos-ribeiro** [comment]
Waving the "new code is covered" criteria for the `OSError` exception since it doesn't seem to aggregate much value to the test suite.

**Comment by MartinBasti** [comment]
> Looks good. Is it possible to test this with integration tests before merging?

Needs config change in test env to enable that plugin

**Comment by MartinBasti** [comment]
Release notes: 
- Add optional integration with [OMPS](https://github.com/release-engineering/operators-manifests-push-service) service to push operator manifest archives automatically into application registry

**Pull request merged**